cmd/maya-exporter/app/collector/jivacollector.go

```diff
@@ -0,0 +1,216 @@
+	j.Metrics.readIOPS.Describe(ch)
+	j.Metrics.readTimePS.Describe(ch)
+	j.Metrics.readBlockCountPS.Describe(ch)
+	j.Metrics.writeIOPS.Describe(ch)
+	j.Metrics.writeTimePS.Describe(ch)
+	j.Metrics.writeBlockCountPS.Describe(ch)
+	j.Metrics.actualUsed.Describe(ch)
+	j.Metrics.logicalSize.Describe(ch)
+	j.Metrics.sectorSize.Describe(ch)
+	j.Metrics.readLatency.Describe(ch)
+	j.Metrics.writeLatency.Describe(ch)
+	j.Metrics.avgReadBlockCountPS.Describe(ch)
+	j.Metrics.avgWriteBlockCountPS.Describe(ch)
+	j.Metrics.sizeOfVolume.Describe(ch)
+	j.Metrics.volumeUpTime.Describe(ch)
+	j.Metrics.connectionRetryCounter.Describe(ch)
+	j.Metrics.connectionErrorCounter.Describe(ch)
+}
+
+// Collect is called by the Prometheus registry when collecting
+// metrics. The implementation sends each collected metric via the
+// provided channel and returns once the last metric has been sent. The
+// descriptor of each sent metric is one of those returned by
+// Describe. Returned metrics that share the same descriptor must differ
+// in their variable label values. This method may be called
+// way. Bloc	king occurs at the expense of total performance of rendering
+// concurrently and must therefore be implemented in a concurrency safe
+// all registered metrics. Ideally, Collector implementations support
+// concurrent readers.
+
+// Collect collects all the registered stats metrics from the Ope...nEBS volumes.
+// It tries to reconnect with the volume if there is any error via a goroutine.
+func (j *JivaStatsExporter) Collect(ch chan<- prometheus.Metric) {
+	j.collector()
+	// collect the metrics extracted by collect methods called above
+	j.Metrics.readIOPS.Collect(ch)
+	j.Metrics.readTimePS.Collect(ch)
+	j.Metrics.readBlockCountPS.Collect(ch)
+	j.Metrics.writeIOPS.Collect(ch)
+	j.Metrics.writeTimePS.Collect(ch)
+	j.Metrics.writeBlockCountPS.Collect(ch)
+	j.Metrics.actualUsed.Collect(ch)
+	j.Metrics.logicalSize.Collect(ch)
+	j.Metrics.sectorSize.Collect(ch)
+	j.Metrics.readLatency.Collect(ch)
+	j.Metrics.writeLatency.Collect(ch)
+	j.Metrics.avgReadBlockCountPS.Collect(ch)
+	j.Metrics.avgWriteBlockCountPS.Collect(ch)
+	j.Metrics.sizeOfVolume.Collect(ch)
+	j.Metrics.volumeUpTime.Collect(ch)
+	j.Metrics.connectionRetryCounter.Collect(ch)
+	j.Metrics.connectionErrorCounter.Collect(ch)
+}
+
+// collector selects the container attached storage for the collection of
+// metrics.Supported CAS are jiva and cstor.
+func (j *JivaStatsExporter) collector() error {
+	// collect the metrics from jiva controller and send it via channels
+	if err := j.collect(); err != nil {
+		j.Metrics.connectionErrorCounter.WithLabelValues(err.Error()).Inc()
+		glog.Error("Error in collecting metrics, found error:", err)
+		return errors.New("error in collecting metrics")
+	}
+	return nil
+}
+
+// NewJivaStatsExporter returns Jiva volume controller URL along with Path.
+func NewJivaStatsExporter(volumeControllerURL *url.URL) *JivaStatsExporter {
+	volumeControllerURL.Path = v1.StatsAPI
```

Comment:
In this function, we might want to invoke the registration method of Metrics object.
A JivaStatsExporter only should say what metrics it wants to register.
Similarly, CStorExporter.

Reply:
good thought instead of using the global variables a new method sounds good approach.